junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/ClassTestDescriptor.java

```diff
@@ -201,11 +205,70 @@ private Object instantiateAndPostProcessTestInstance(JupiterEngineExecutionConte
 		return instance;
 	}
 
+	protected Object invokeTestInstanceConstructor(Object outerInstance, ExtensionRegistry registry,
+			ExtensionContext extensionContext) {
+		Constructor<?> constructor = ReflectionUtils.getDeclaredConstructor(this.testClass);
+		if (outerInstance != null) {
+			return executableInvoker.invoke(constructor, outerInstance, extensionContext, registry);
+		}
+		else {
+			return executableInvoker.invoke(constructor, extensionContext, registry);
+		}
+	}
+
+	protected TestInstanceFactory resolveTestInstanceFactory(ExtensionRegistry registry,
+			ExtensionRegistry parentRegistry) {
+
+		List<TestInstanceFactory> factories = registry.getExtensions(TestInstanceFactory.class);
+		if (factories.isEmpty()) {
+			return null;
+		}
+
+		if (parentRegistry != null) {
+			List<TestInstanceFactory> parentFactories = parentRegistry.getExtensions(TestInstanceFactory.class);
+			factories.removeAll(parentFactories);
+		}
+
+		if (factories.size() > 1) {
+			String factoryNames = factories.stream().map(factory -> factory.getClass().getSimpleName()).collect(
+				Collectors.joining(","));
+
+			String errorMessage = String.format(
+				"Too many TestInstanceFactory extensions [%s] registered on test class: %s", factoryNames,
+				testClass.getSimpleName());
+
+			throw new ExtensionContextException(errorMessage);
+		}
+
+		return factories.get(0);
+	}
+
+	protected Object invokeTestInstanceFactory(Object outerInstance, ExtensionRegistry registry,
+			ExtensionRegistry parentRegistry, ExtensionContext extensionContext) {
+
+		TestInstanceFactory factory = resolveTestInstanceFactory(registry, parentRegistry);
+		if (factory == null) {
+			return invokeTestInstanceConstructor(outerInstance, registry, extensionContext);
+		}
+
+		try {
+			if (outerInstance != null) {
+				return factory.instantiateNestedTestClass(this.testClass, outerInstance, extensionContext);
```

Comment:
This approach makes it impossible for such a constructor to accept arguments provided by a `ParameterResolver` -- for example, `TestInfo`, `TestReporter`, mocks from Mockito, beans from Spring, etc.

We need to decide if we want this new extension API to suffer from this restriction.

Reply:
I do not believe this is a restriction that impacts the majority of use cases for factories. I believe factories are specifically for when customized construction is required by external libraries; in which case arguments from JUnit 5 sources are probably not going to be required. Also, the current extension API seems to cover these situations well and using a factory doesn't negate post-processing or any other extension.